scripts/AbinsModules/Instruments/Broadening.py

```diff
@@ -0,0 +1,575 @@
+    :type frequencies: 1D array-like
+    :param bins:
+        Evenly-spaced frequency bin values for the output spectrum. If *frequencies* is None, s_dft is assumed to
+        correspond to mid-point frequencies on this mesh.
+    :type bins: 1D array-like
+    :param s_dft: scattering values corresponding to *frequencies*
+    :type s_dft: 1D array-like
+    :param sigma:
+        width of broadening function. This may be a scalar used over the whole spectrum, or a series of values
+        corresponding to *frequencies*.
+    :type sigma: float or 1D array-like
+    :param scheme:
+        Name of broadening method used. Options:
+
+        - none: Return the input data
+        - gaussian: Evaluate a Gaussian on the output grid for every input point and sum them. Simple but slow, and
+              recommended only for benchmarking and reference calculations.
+        - normal: Generate histograms with appropriately-located normal distributions for every input point. In
+              principle this is more accurate than 'Gaussian' but in practice results are very similar and just as slow.
+        - gaussian_truncated: Gaussians are cut off at a range of 3 * max(sigma). This leads to small discontinuities in
+              the tails of the broadened peaks, but is _much_ faster to evaluate as all the broadening functions are the
+              same (limited) size. Recommended for production calculations.
+        - normal_truncated: As gaussian_truncated, but with normally-distributed histograms replacing Gaussian evaluation.
+              Results are marginally more accurate at a marginally increased cost.
+        - interpolate: A possibly-novel approach in which the spectrum broadened with a frequency-dependent kernel is
+              approximated by interpolation between a small set of spectra broadened with logarithmically-spaced
+              fixed-width kernels. In principle this method introduces an incorrect asymmetry to the peaks. In practice
+              the results are visually acceptable for a slowly-varying instrumental broadening function, with greatly
+              reduced computation time. The default spacing factor of sqrt(2) gives error of ~1%.
+        - interpolate_coarse: The approximate interpolative scheme (above) with a spacing factor of 2, which yields
+              error of ~5%. This is more likely to cause artefacts in the results... but it is very fast. Not recomended
+              for production calculations.
+
+    :type scheme: str
+
+    :returns: (freq_points, broadened_spectrum)
+
+    The *freq_points* are the mid-bin frequency values which
+    nominally correspond to *broadened_spectrum*; both of these are
+    1D arrays of length (N-1) corresponding to the length-N *bins*.
+    """
+    if (bins is None) or (s_dft is None) or (sigma is None):
+        raise ValueError("Frequency bins, S data and broadening width must be provided.")
```

Comment:
That does seem more reasonable. Abins in general does use a lot of keywords for essential arguments, I must have missed that somewhere.

Actually `sigma=None` should be allowed anyway, as it isn't used by the null algorithm.

Reply:
Going to avoid setting a default that breaks most cases though. Instead will add a docstring note that 'none' ignores sigma, and continue to demand that Sigma is numerical.

Type hinting would be quite handy here, is there much interest in using that in Mantid?

Reply:
Once we are fully python3 there is quite a lot of interest in doing type hinting and potentially running something like [mypy](https://pypi.org/project/mypy/) to the build servers to check it. I know some people are already using type hinting and just planning to hold off on mergeing the PR's until we are fully python3 which should be soonish.

So in summary if you want to add some type hinting and don't mind waiting a bit before this gets merged I would say go for it.

Reply:
I do mind waiting in this case so will hold off. Once the release is out I might start using it.
One problem is that I believe type hinting is still quite limited for numpy objects.